### PR TITLE
PXP-489 retry request on query timeout

### PR DESCRIPTION
**Pull request by Fadhil**
## Summary

In this PR, we introduce a `retry` mechanism for requests that `timeout`.  If a request to our third-party APIs timeout, the request is retried up to 3 times (waiting 5 seconds after each attempt).

xtTicket: https://mindvalley.atlassian.net/browse/PXP-489

## What's Changed

- [x] `QueryClient.call_api` to retry requests if they timeout
- [x] Retry maximum of 3 times
- [x] Backoff 5 seconds for each retry

Screenshot 2023-03-21 at 3 51 58 PM
Screenshot 2023-03-21 at 3 29 12 PM

### CODE REVIEW

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/77#" title="Author self-approved">Fadhil</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: src/error.rs
Comment (jk-gan) [inline src/error.rs]:
```suggestion
```
I think we don't need this since we  got them under `APIError`

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:115]: I think we don't have to pass `client` here, we can get the client value from `self`

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:103]: need to put back debug log for request

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:149]: maybe we can show more in the debug log here: 
```rust
debug!("{domain} is timeout");
```

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/graphql/mod.rs
Comment (Fadhil) [inline src/graphql/mod.rs:150]: Do we still need a debug here if we're printing an error?

Comment (jk-gan) [reply src/graphql/mod.rs:150]: yes if you think a debug log can show more context

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:156]: can you use the prettier debug format here? 

`debug!("request: {:#?}", request);`

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:126]: the first api request doesn't have debug log

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/graphql/mod.rs
Comment (Fadhil) [reply src/graphql/mod.rs:150]: In this case, the error printed out is the same as the debug, and doesn't seem to add anything. Shall we remove this?

Comment (jk-gan) [reply src/graphql/mod.rs]: ya I think we can remove it

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:128]: typo here

Comment (jk-gan) [reply src/graphql/mod.rs:128]: and here say `GraphQL request`, but below say `request`, I think we should make them consistent

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/graphql/mod.rs
Comment (jk-gan) [inline src/graphql/mod.rs:107]: I think we don't need a debug log here for `response` anymore, we already cover all of them inside `retry_request`

**Review by jk-gan — APPROVED** [review]

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/77#" title="Author self-approved">Fadhil</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/77#pullrequestreview-1364263485" title="Approved">jk-gan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by onimsha — APPROVED** [review]

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/77#" title="Author self-approved">Fadhil</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/77#pullrequestreview-1364263485" title="Approved">jk-gan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/77#pullrequestreview-1364410771" title="Approved">onimsha</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
